modules/navier_stokes/src/fv/kernels/NSFVBase.C

```diff
@@ -0,0 +1,153 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "NSFVBase.h"
+
+#ifdef MOOSE_GLOBAL_AD_INDEXING
+
+#include "InputParameters.h"
+#include "SubProblem.h"
+#include "MooseVariableFV.h"
+
+std::vector<std::unordered_map<const Elem *, ADReal>> NSFVBase::_rc_a_coeffs;
+
+InputParameters
+NSFVBase::validParams()
+{
+  InputParameters params = emptyInputParameters();
+  params.addRequiredCoupledVar("pressure", "The pressure variable.");
+  params.addRequiredCoupledVar("u", "The velocity in the x direction.");
+  params.addCoupledVar("v", "The velocity in the y direction.");
+  params.addCoupledVar("w", "The velocity in the z direction.");
+
+  MooseEnum velocity_interp_method("average rc", "rc");
+
+  params.addParam<MooseEnum>(
+      "velocity_interp_method",
+      velocity_interp_method,
+      "The interpolation to use for the velocity. Options are "
+      "'average' and 'rc' which stands for Rhie-Chow. The default is Rhie-Chow.");
+
+  params.addRequiredParam<Real>("mu", "The viscosity");
+  params.addRequiredParam<Real>("rho", "The density");
+  return params;
+}
+
+NSFVBase::NSFVBase(const InputParameters & params)
+  : _nsfv_subproblem(*params.getCheckedPointerParam<SubProblem *>("_subproblem")),
+    _nsfv_tid(params.get<THREAD_ID>("_tid")),
+    _p_var(dynamic_cast<const MooseVariableFV<Real> *>(&_nsfv_subproblem.getVariable(
+        _nsfv_tid, params.get<std::vector<VariableName>>("pressure").front()))),
+    _u_var(dynamic_cast<const MooseVariableFV<Real> *>(&_nsfv_subproblem.getVariable(
+        _nsfv_tid, params.get<std::vector<VariableName>>("u").front()))),
+    _v_var(params.isParamValid("v")
+               ? dynamic_cast<const MooseVariableFV<Real> *>(&_nsfv_subproblem.getVariable(
+                     _nsfv_tid, params.get<std::vector<VariableName>>("v").front()))
+               : nullptr),
+    _w_var(params.isParamValid("w")
+               ? dynamic_cast<const MooseVariableFV<Real> *>(&_nsfv_subproblem.getVariable(
+                     _nsfv_tid, params.get<std::vector<VariableName>>("w").front()))
+               : nullptr),
+    _mu(params.get<Real>("mu")),
+    _rho(params.get<Real>("rho"))
```

Comment:
You decided against the material property approach?  That does make things more limited w.r.t. dealing with different formulations (conservative vs primitive, effective velocity, etc.).  What was you reasoning behind this?  Especially since the (eventually) coming CNSFV will work that way.

Reply:
I was punting on the fact that `NSFVBase` does not inherit from `MaterialPropertyInterface` (and cannot because of the diamond problem).

To keep code duplication down, I can in `NSFVBase` either:
- Implement `mu` and `rho` virtual functions
- Implement templated functions that accept `mu` and `rho` functors as arguments

I'm leaning towards the latter

Reply:
Jk, I'm working on just adding `mu` and `rho` parameters to the `rcCoeff` API

Reply:
Sorry - I didn't just mean mu and rho.  I primarily meand p, u, v, w, rho - the flow variables.  The CNSFV gets all of them as material properties.  This allows users to work with them in input files in any combination of p, rho, T, E, v, rhov, vol_frac*rhov, etc. - without any fancy changes to kernels, bcs, etc - which is why we did things like that in Pronghorn.  This breaks that convention.  If we keep the compressible stuff from there ~like it is, then the user experience and capability will be inconsistent with this stuff.

Reply:
Pasting from slack:

> Ok I will try to update that. This one will be harder because yea now I won't be able to call Variable methods.
Or maybe I just need to make this INSFVAdvectionBase instead of NSFVAdvectionBase because those variable method calls are only done for Rhie-Chow interpolations which are only done for INSFV

Reply:
`rho` is now a constant for `INSFV` per our discussions